lte/gateway/c/oai/tasks/mme_app/mme_app_ip_imsi.cpp

```diff
@@ -0,0 +1,96 @@
+#include <bits/stdc++.h>
+#include <unordered_map>
+#include <iostream>
+#include "mme_app_ip_imsi.h"
+using namespace std;
+typedef unordered_map<string, vector<uint64_t>> Ipv4Map;
+
+Ipv4Map ipv4map;
+
+void initialize_ipv4_map() {
+  OAILOG_FUNC_IN(LOG_MME_APP);
+  ipv4map = Ipv4Map{};
+  OAILOG_FUNC_OUT(LOG_MME_APP);
+}
+
+void mme_app_log_ipv4_imsi_map() {
+  OAILOG_FUNC_IN(LOG_MME_APP);
+  for (auto itr = ipv4map.begin(); itr != ipv4map.end(); ++itr) {
+    for (auto it_vec = itr->second.begin(); it_vec != itr->second.end();
+         ++it_vec) {
+      OAILOG_TRACE(
+          LOG_MME_APP, "ue_ip: %s \t imsi:%lu \n", itr->first.c_str(), *it_vec);
+    }
+    OAILOG_TRACE(LOG_MME_APP, "\n");
+  }
+  OAILOG_FUNC_OUT(LOG_MME_APP);
+}
+
+int mme_app_insert_ue_ipv4_addr(uint32_t ipv4_addr, uint64_t imsi64) {
+  OAILOG_FUNC_IN(LOG_MME_APP);
+  char ipv4[INET_ADDRSTRLEN] = {0};
+  inet_ntop(AF_INET, (void*) &ipv4_addr, ipv4, INET_ADDRSTRLEN);
+  auto itr = ipv4map.find(ipv4);
+  if (itr == ipv4map.end()) {
+    vector<uint64_t> vec = {};
+    vec.insert(vec.begin(), imsi64);
+    ipv4map[ipv4] = vec;
+    OAILOG_DEBUG_UE(LOG_MME_APP, imsi64, "Inserting ue_ip:%x \n", ipv4_addr);
+  } else {
+    OAILOG_DEBUG_UE(
+        LOG_MME_APP, imsi64, "Inserting imsi for existing ue_ip:%x \n",
+        ipv4_addr);
+    ipv4map[ipv4].push_back(imsi64);
+  }
+  OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNok);
+}
+
+int mme_app_get_imsi_from_ipv4(uint32_t ipv4_addr, uint64_t* imsi64) {
```

Comment:
with your current usage, imsi64 is an array of size 2. this function will go out of bound if there are more than 2 entries for a given IP addr (think about multi PGW case). It is better to calloc imsi64 in this function and return size rather than RETURNOk, RETURNerror, etc. Of course do not forget to free the memory by the caller.